test/parallel_api/iterator/tuple_unit.pass.cpp

```diff
@@ -0,0 +1,191 @@
+    EXPECT_EQ((t1 < t2), (onedpl_t1 < onedpl_t2), "less than comparison does not match std::tuple");
+    EXPECT_EQ((t1 <= t2), (onedpl_t1 <= onedpl_t2), "less than or equal to comparison does not match std::tuple");
+    EXPECT_EQ((t1 > t2), (onedpl_t1 > onedpl_t2), "greater than comparison does not match std::tuple");
+    EXPECT_EQ((t1 >= t2), (onedpl_t1 >= onedpl_t2), "greater than or equal to comparison does not match std::tuple");
+
+    EXPECT_EQ((t1 == t2), (t1 == onedpl_t2), "equality comparison does not match std::tuple");
+    EXPECT_EQ((t1 != t2), (t1 != onedpl_t2), "inquality comparison does not match std::tuple");
+    EXPECT_EQ((t1 < t2), (t1 < onedpl_t2), "less than comparison does not match std::tuple");
+    EXPECT_EQ((t1 <= t2), (t1 <= onedpl_t2), "less than or equal to comparison does not match std::tuple");
+    EXPECT_EQ((t1 > t2), (t1 > onedpl_t2), "greater than comparison does not match std::tuple");
+    EXPECT_EQ((t1 >= t2), (t1 >= onedpl_t2), "greater than or equal to comparison does not match std::tuple");
+
+    EXPECT_EQ((t1 == t2), (onedpl_t1 == t2), "equality comparison does not match std::tuple");
+    EXPECT_EQ((t1 != t2), (onedpl_t1 != t2), "inquality comparison does not match std::tuple");
+    EXPECT_EQ((t1 < t2), (onedpl_t1 < t2), "less than comparison does not match std::tuple");
+    EXPECT_EQ((t1 <= t2), (onedpl_t1 <= t2), "less than or equal to comparison does not match std::tuple");
+    EXPECT_EQ((t1 > t2), (onedpl_t1 > t2), "greater than comparison does not match std::tuple");
+    EXPECT_EQ((t1 >= t2), (onedpl_t1 >= t2), "greater than or equal to comparison does not match std::tuple");
+
+    auto onedpl_t3 = to_onedpl_tuple(t1);
+    auto onedpl_t4 = to_onedpl_tuple(t2);
+    auto t3 = t1;
+    auto t4 = t2;
+
+    t4 = onedpl_t1;
+    EXPECT_TRUE(t1 == t4, "assignment of oneDPL tuple to std::tuple provides incorrect results");
+
+    t3 = onedpl_t2;
+    EXPECT_TRUE(t2 == t3, "assignment of oneDPL tuple to std::tuple provides incorrect results");
+
+    onedpl_t3 = t1;
+    EXPECT_TRUE(onedpl_t1 == onedpl_t3, "assignment of oneDPL tuple from std::tuple provides incorrect results");
+
+    onedpl_t4 = t2;
+    EXPECT_TRUE(onedpl_t2 == onedpl_t4, "assignment of oneDPL tuple from std::tuple provides incorrect results");
+
+    decltype(onedpl_t1) onedpl_t5 = onedpl_t2;
+    decltype(onedpl_t1) onedpl_t6 = onedpl_t1;
+
+    swap(onedpl_t5, onedpl_t6);
+    EXPECT_TRUE(((onedpl_t1 == onedpl_t5) && (onedpl_t6 == onedpl_t2)),
+                "swap of oneDPL tuple provides incorrect results");
+}
+
+constexpr void
+test_compile_time()
+{
+    constexpr std::tuple<int, int, uint64_t> std_tup1{1, 2, 3};
+    constexpr std::tuple<int, uint32_t, int> std_tup2{99, 2, 3};
+    constexpr oneapi::dpl::tuple<int, int, uint64_t> dpl_tup1{1, 2, 3};
+    constexpr oneapi::dpl::tuple<int, uint32_t, int> dpl_tup2{99, 2, 3};
```

Comment:
Yes, this is a sizeable mistake, good catch.  Changing to internal tuple reveals many missing elements to handle `constexpr` properly.  None are complicated (adding `constexpr` to methods), but they need to be fixed to support this functionality.

Reply:
Resolved the `constexpr` additions.  If we don't want to add these in this PR, we can strip out the constexpr changes, and this part of the testing.  I don't have a strong opinion here, it is not needed for the use case being requested.

Reply:
The additions are not huge, but they seem kind of strange comparing to other code. I do not think we have / will ever have use cases for constexpr operations with the internal tuple. Do you have some scenario in mind where it could be useful? Doing it "just in case" does not sound justified.

Reply:
My intention was just to get our internal tuple and its support for the binary comparison operators closer to `std::tuple` specification.  I don't have a specific use case in mind, and such a case probably does not exist. A user should only encounter these as a result of `zip_iterator` which don't have the ability to create tuples at compile time.

After consideration, I think it is probably best to remove support and the testing of `constexpr` part of this.  We can always add it in the future if we have a reason to do so.